compiler/damlc/tests/src/DA/Test/Packaging.hs

```diff
@@ -780,7 +780,8 @@ tests Tools{damlc} = testGroup "Packaging" $
             ]
           writeFileUTF8 (projDir </> "main" </> "Main.daml") $ unlines
             [ "module Main where"
-            , "import \"dependency\" B"
+            -- , "import \"dependency\" B" -- this brings into scope the `C` for which we don't have an instance
+            , "import \"data-dependency\" B"
```

Comment:
One of the things DataDependencies.hs tries to do is to rewrite typeclass references from data-dependencies to something in the dependencies, if it happens to share the same module, name and definition (down to the types). This is taking advantage of the fact that in Daml-LF typeclasses (type synonyms + structs) are structural, while in Daml typeclasses are nominal.

The advantage of this is for upgrading. It makes interop easier because you're not dealing with multiple typeclasses that differ only in which package they come from. So you can e.g. make calls to functions in a data-dependency using an instance from your dependency, or vice versa. I think this is worth preserving if possible. 

What's happening here, I think, is that by changing the order of packages getting registered in the package db, you are preventing DataDependencies.hs from rewriting the typeclass, since it doesn't know about the dependency. What would be nice is if all dependencies were written as early as possible (taking into account any data-dependencies they may require). Not sure how to do that though. :-|

Reply:
So the testcase was added in https://github.com/digital-asset/daml/pull/5330 and it has an explanation of what we’re trying to test for here. Specifically the crucial part is that when we register the data-dependency the dependency is already in scope. So I think you’ve lost the meaning of the test here. The dependency is now no longer registered before so if we break exposed modules again nothing actually breaks.

I see two options for fixing this:
1. If we have non-exposed modules in daml-prim & daml-stdlib, you could probably use those but I think we don’t unfortunately.
2. If that doesn’t work, maybe we can actually make the data-dependency depend on the dependency. This seems a bit dumb but I think it might recover the goal of the test.

As for what @sofiafaro-da is mentioning, I think the analysis is exactly right. However, I’m not completely sure we want to preserve this behavior. Initially, this was driven by daml-prim & daml-stdlib. We cannot have stable package ids to the dependencies there reasonably easily since GHC really treats those magically. So we get a new Eq instance and we want to treat those identical to the old Eq instance.
For a while, we tried to push people towards defining non-template packages for their typeclasses that you can freely recompile and always use as a dependency where this behavior also made sense for those packages. However, that didn’t really play out in practice and instead we added better support for importing typeclasses themselves as data-dependencies by supporting functional dependencies, minimal pragmas, …. and afaik that’s what everyone is using.

So at this point I’d say I’d discourage having multiple definitions of your own typeclasses in your dependency graph and instead reference those via data-dependencies themselves so they get stable pkg ids. we should check that with product but I highly doubt you’ll get any pushback.

Alternatively to recover the previous behavior while also fixing the issue you can make a special case for deps that depend only on other deps & stable packages and register all of those first. That’s the only thing that this worked before anyway.

Reply:
For the time being, I ended up taking the approach you suggest at the end, namely, the subset of regular dependencies that only depend on on other regular dependencies, stable packages and builtin packages (call it `r`) takes precedence on the dependency graph. This is achieved by extending the proper dependency graph with edges from each data-dependency to each element of `r`.